community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java

```diff
@@ -45,141 +41,202 @@
+        value.map( dryRunMapper );
+        if ( dryRunMapper.needsConversion )
+        {
+            WritingMaterializeValueMapper realMapper = new WritingMaterializeValueMapper( proxySPI );
+            return value.map( realMapper );
+        }
+        return value;
     }
 
-    public static AnyValue mapAnyValue( EmbeddedProxySPI proxySPI, AnyValue value )
+    private static final class WritingMaterializeValueMapper extends AbstractMaterializeValueMapper
     {
-        CompiledMaterializeValueMapper mapper = new CompiledMaterializeValueMapper( proxySPI );
-        return value.map( mapper );
-    }
+        private EmbeddedProxySPI proxySpi;
 
-    // Create proxy wrapping values for nodes, relationships and paths that are not already such values
+        WritingMaterializeValueMapper( EmbeddedProxySPI proxySpi )
+        {
+            this.proxySpi = proxySpi;
+        }
 
-    @Override
-    public AnyValue mapNode( VirtualNodeValue value )
-    {
-        if ( value instanceof NodeProxyWrappingNodeValue )
+        // Create proxy wrapping values for nodes and relationships that are not already such values
+
+        @Override
+        public AnyValue mapNode( VirtualNodeValue value )
         {
-            return value;
+            if ( value instanceof NodeValue )
+            {
+                return value;
+            }
+            return ValueUtils.fromNodeProxy( proxySpi.newNodeProxy( value.id() ) );
+        }
+
+        @Override
+        public AnyValue mapRelationship( VirtualRelationshipValue value )
+        {
+            if ( value instanceof RelationshipValue )
+            {
+                return value;
+            }
+            return ValueUtils.fromRelationshipProxy( proxySpi.newRelationshipProxy( value.id() ) );
+        }
+
+        // Recurse through maps and sequences
+
+        @Override
+        public AnyValue mapMap( MapValue value )
+        {
+            Map<String,AnyValue> map = new HashMap<>();
+            value.foreach( ( k, v ) -> map.put( k, v.map( this ) ) );
+            return VirtualValues.map( map );
+        }
+
+        @Override
+        public AnyValue mapSequence( SequenceValue value )
+        {
+            List<AnyValue> list = new ArrayList<>( value.length() );
+            value.forEach( v -> list.add( v.map( this ) ) );
+            return VirtualValues.fromList( list );
         }
-        return ValueUtils.fromNodeProxy( defaultValueMapper.mapNode( value ) );
     }
 
-    @Override
-    public AnyValue mapRelationship( VirtualRelationshipValue value )
+    private static final class DryRunMaterializeValueMapper extends AbstractMaterializeValueMapper
     {
-        if ( value instanceof RelationshipProxyWrappingValue )
+        boolean needsConversion;
+
+        @Override
+        public AnyValue mapNode( VirtualNodeValue value )
         {
+            if ( value instanceof NodeValue )
```

Comment:
Maybe
```
{
    needsConversion = !value && value instanceof NodeReference;
    return value;
}
```

Reply:
I think you mean `needsConversion || (value instanceof NodeReference)`